File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
from collections import deque

input = sys.stdin.readline

N = int(input())

visited = [[True] * N for _ in range(N)]
graph = [[0] * N for _ in range(N)]

dir = [(0, 1), (0, -1), (1, 0), (-1, 0)]

for row in range(N):
  string = input().rstrip()
  col = 0
  for s in string:
    if s == "1":
      visited[row][col] = False
      graph[row][col] = 1
    col += 1

houseCount = []

for row in range(N):
  for col in range(N):
    if visited[row][col]:
      continue
    
    count = 1
    q = deque()
    q.append((row, col))
    visited[row][col] = True

    while q:
      cur_row, cur_col = q.popleft()

      for r, c in dir:
        next_row = cur_row + r
        next_col = cur_col + c

        if next_row >= 0 and next_row < N and next_col >= 0 and next_col < N:
          if not visited[next_row][next_col]:
            visited[next_row][next_col] = True
            q.append((next_row, next_col))
            count += 1

    houseCount.append(count)

houseCount.sort()
print(len(houseCount))
for c in houseCount:
  print(c)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.swift
import Foundation

let n = Int(readLine()!)!

var graph: [[Int]] = []
var visited: [[Bool]] = Array(
  repeating: Array(repeating: false, count: n), 
  count: n
)
var drdc = [(0, 1), (0, -1), (1, 0), (-1, 0)]

var result: [Int] = []

for _ in 0..<n {
  let input = readLine()!.map { Int(String($0))! }
  graph.append(input)
}

for row in 0..<n {
  for col in 0..<n {
    if graph[row][col] == 0 { continue }
    if visited[row][col] { continue }

    var queue: [(Int, Int)] = []
    var front = 0
    
    visited[row][col] = true

    queue.append((row, col))
    var count = 0

    while queue.count >= front + 1 {
      let (curR, curC) = queue[front]
      front += 1
      count += 1

      for (r, c) in drdc {
        let nextR = curR + r
        let nextC = curC + c

        if nextR >= 0 && nextR < n && nextC >= 0 && nextC < n {
          if graph[nextR][nextC] != 0 && !visited[nextR][nextC] {
            queue.append((nextR, nextC))
            visited[nextR][nextC] = true
          }
        }
      }
    }

    result.append(count)
  }
}

print(result.count)
result.sort()
for i in 0..<result.count {
  print(result[i])
}
